api/inc/register_gateway.h

```diff
@@ -17,105 +17,146 @@
- * ARM Thumb (w/o the Thumb-2 extension): 0xF7FxAxxx (32 bits)
- * The 'x's can be freely chosen (they have been chosen randomly here).
- * This requires the Thumb-2 extensions because otherwise only 16 bits opcodes
- * are accepted.*/
-#if defined(__thumb__) && defined(__thumb2__)
-#define UVISOR_REGISTER_GATEWAY_MAGIC 0xF7F3A89E
-#else
-#error "Unsupported instruction set. The ARM Thumb-2 instruction set must be supported."
-#endif /* __thumb__ && __thumb2__ */
-
-/* Register gateway operations */
-/* Note: Do not use special characters as these numbers will be stringified. */
-#define UVISOR_OP_READ(op)  (op)
-#define UVISOR_OP_WRITE(op) ((1 << 4) | (op))
-#define UVISOR_OP_NOP       0x0
-#define UVISOR_OP_AND       0x1
-#define UVISOR_OP_OR        0x2
-#define UVISOR_OP_XOR       0x3
-
-/* Default mask for whole register operatins. */
-#define __UVISOR_OP_DEFAULT_MASK 0x0
-
-/* Register gateway metadata */
-#if defined(__CC_ARM)
-
-/* TODO/FIXME */
-
-#elif defined(__GNUC__)
-
-/* 1 argument: Simple read, no mask */
-#define __UVISOR_REGISTER_GATEWAY_METADATA1(src_box, addr) \
-    "b.n skip_args%=\n" \
-    ".word " UVISOR_TO_STRING(UVISOR_REGISTER_GATEWAY_MAGIC) "\n" \
-    ".word " UVISOR_TO_STRING(addr) "\n" \
-    ".word " UVISOR_TO_STRING(src_box) "_cfg_ptr\n" \
-    ".word " UVISOR_TO_STRING(UVISOR_OP_READ(UVISOR_OP_NOP)) "\n" \
-    ".word " UVISOR_TO_STRING(__UVISOR_OP_DEFAULT_MASK) "\n" \
-    "skip_args%=:\n"
-
-/* 2 arguments: Simple write, no mask */
-#define __UVISOR_REGISTER_GATEWAY_METADATA2(src_box, addr, val) \
-    "b.n skip_args%=\n" \
-    ".word " UVISOR_TO_STRING(UVISOR_REGISTER_GATEWAY_MAGIC) "\n" \
-    ".word " UVISOR_TO_STRING(addr) "\n" \
-    ".word " UVISOR_TO_STRING(src_box) "_cfg_ptr\n" \
-    ".word " UVISOR_TO_STRING(UVISOR_OP_WRITE(UVISOR_OP_NOP)) "\n" \
-    ".word " UVISOR_TO_STRING(__UVISOR_OP_DEFAULT_MASK) "\n" \
-    "skip_args%=:\n"
-
-/* 3 arguments: Masked read */
-#define __UVISOR_REGISTER_GATEWAY_METADATA3(src_box, addr, op, mask) \
-    "b.n skip_args%=\n" \
-    ".word " UVISOR_TO_STRING(UVISOR_REGISTER_GATEWAY_MAGIC) "\n" \
-    ".word " UVISOR_TO_STRING(addr) "\n" \
-    ".word " UVISOR_TO_STRING(src_box) "_cfg_ptr\n" \
-    ".word " UVISOR_TO_STRING(UVISOR_OP_READ(op)) "\n" \
-    ".word " UVISOR_TO_STRING(mask) "\n" \
-    "skip_args%=:\n"
-
-/* 4 arguments: Masked write */
-#define __UVISOR_REGISTER_GATEWAY_METADATA4(src_box, addr, val, op, mask) \
-    "b.n skip_args%=\n" \
-    ".word " UVISOR_TO_STRING(UVISOR_REGISTER_GATEWAY_MAGIC) "\n" \
-    ".word " UVISOR_TO_STRING(addr) "\n" \
-    ".word " UVISOR_TO_STRING(src_box) "_cfg_ptr\n" \
-    ".word " UVISOR_TO_STRING(UVISOR_OP_WRITE(op)) "\n" \
-    ".word " UVISOR_TO_STRING(mask) "\n" \
-    "skip_args%=:\n"
-
-#endif /* __CC_ARM or __GNUC__ */
+/** Both `uvisor_read` and `uvisor_write` can have 2 additional input arguments -
+ *  mask and operation. These macros select the right argument from
```

Comment:
I'd rather not we have optional macro arguments, for code maintainability sake. We need to balance API ease of understanding and use with code maintenance and future refactoring efforts.

I'd prefer named macros like `uvisor_write_mask` that takes a mask and `uvisor_write_op` that takes an op, and `uvisor_write_mask_op` that takes  both. The macros that accept fewer parameters can be implemented in terms of the macros that take more parameters; this maintains the aspect of your current implementation where the default values for parameters easy to identify.